From a8d843e70b51a327f5e2879c27bf62cf275c63b7 Mon Sep 17 00:00:00 2001
From: Amith A <amitajit@qti.qualcomm.com>
Date: Mon, 10 Nov 2025 20:01:23 +0530
Subject: [PATCH 1/3] QSDK:PROTOCOL:hostapd: Add QCA vendor attr for Estimated
 Service Parameter (ESP) feature

Add new vendor attributes for Estimated Service Parameter (ESP) related
parameters such as airtime fraction, PPDU duration, Block Ack (BA) window
size, and ESP enable status. These attributes can be used to set or get
values of ESP parameters in driver/firmware from userspace.

Patch-dependency: None

Signed-off-by: Amith A <amitajit@qti.qualcomm.com>
---
 src/common/qca-vendor.h | 43 +++++++++++++++++++++++++++++++++++++++++
 1 file changed, 43 insertions(+)

--- a/src/common/qca-vendor.h
+++ b/src/common/qca-vendor.h
@@ -3907,6 +3907,12 @@ enum qca_wlan_vendor_attr_config {
 	 */
 	QCA_WLAN_VENDOR_ATTR_CONFIG_RADIO_INDEX = 135,
 
+	/* Nested attribute to configure Estimated Service Parameters (ESP) to
+	 * the firmware. This attribute contains nested attributes defined in
+	 * enum qca_wlan_vendor_attr_config_esp_param. This attribute is also
+	 * used to retrieve the parameters from firmware.
+	 */
+	QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_PARAMS = 142,
 
 	/* keep last */
 	QCA_WLAN_VENDOR_ATTR_CONFIG_AFTER_LAST,
@@ -19237,4 +19243,46 @@ enum qca_wlan_vendor_atf_offload_operati
 	QCA_WLAN_VENDOR_ATF_OFFLOAD_STATS = 11,
 };
 
+/**
+ * enum qca_wlan_vendor_attr_config_esp_param - Parameters for ESP configuration
+ *
+ * This enum defines the attributes used to configure and retrieve
+ * Estimated Service Parameters (ESP) settings.
+ *
+ * @QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_ENABLE: u8 attribute. Enable or disable ESP
+ * functionality. Set to 1 to enable ESP, 0 to disable. When enabled,
+ * the ESP parameters (airtime, PPDU duration, BA window) are advertised in
+ * management frames.
+ *
+ * @QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_AIRTIME: u32 attribute. Represents the
+ * predicted percentage of airtime that a new STA joining the BSS can expect
+ * to be available for the transmission. The lower 8 bits represent the airtime
+ * of BE AC, linearly scaled with 255 representing 100% and 0 representing 0%.
+ * Bits 8-15 represent BK AC, bits 16-23 represent VI AC, and bits 24-31
+ * represent VO AC (if applicable). When used to report airtime of all ACs,
+ * each byte represents one AC.
+ *
+ * @QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_PPDU_DUR: u8 attribute. Indicates the
+ * expected target duration of PPDUs transmitted to the STA, in microseconds.
+ * Valid range: 0-255 microseconds.
+ *
+ * @QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_BA_WINDOW: u8 attribute. Specifies the
+ * Block Ack (BA) window size. Valid range: 0-7, where 0 represents Block
+ * Ack not expected to be used, and rest of the values representing how many
+ * MPDUs can be acknowledged in one Block Ack frame. Value 1 for 2 MPDUs,
+ * value 2 for 4 MPDUs and so on up to the value 7 for 64 MPDUs.
+ */
+enum qca_wlan_vendor_attr_config_esp_param {
+	QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_INVALID = 0,
+	QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_ENABLE = 1,
+	QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_AIRTIME = 2,
+	QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_PPDU_DUR = 3,
+	QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_BA_WINDOW = 4,
+
+	/* keep last */
+	QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_AFTER_LAST,
+	QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_MAX =
+		QCA_WLAN_VENDOR_ATTR_CONFIG_ESP_AFTER_LAST - 1,
+};
+
 #endif /* QCA_VENDOR_H */
